### PR TITLE
Add MC_RESERVE_CHANNEL_IDS frame

### LINKED ISSUE #9: Add something like a MC_RESERVE_SESSIONIDS frame.
From Max in issue #4 comment:
https://github.com/GrumpyOldTroll/draft-jholland-quic-multicast/pull/8#discussion_r854656985

There should probably be a mechanic to force clients to not use some connection IDs for unicast connections. I think only doing it in the MC_SESSION_PROPERTIES frame is too late, it should probably be done immediately after the handshake of the unicast connection. Its basically a frame telling the client "Do not use any of these connection IDs for your unicast streams as we might have a multicast session that uses this ID. If you (by sheer bad luck) already use one of them for a unicast connection (i.e. the initially created one), issue a new connection ID and retire the old one immediately."

### PR DESCRIPTION
**Pull request by MaxF12**
Closes #9

### CODE REVIEW

**Comment by MaxF12** [comment]
This might need a second list with the lengths of the channel IDs, which seemed a bit awkward but I guess it’s necessary?

File: draft-jholland-quic-multicast.md [hunk 1]
```diff
@@ -221,6 +221,34 @@ If a server or client somehow still detect a stateless reset for a channel, they
...
+If a server learns that a client supports multicast through its transport parameters, it SHOULD immediately send a MC_RESERVE_CHANNEL_IDS frame (type=TBD-00) that includes all channel IDs that might be used by its multicast channels.
+This is done to prevent a client from using those connection IDs for the unicast connection, which could cause a delay when attempting to join a channel as they would first have to be retired for the unicast connection.
+
+MC_RESERVE_CHANNEL_IDS frames are formatted as shown in {{fig-mc-channel-reserve-ids}}.
+
+~~~
+MC_RESERVE_CHANNEL_IDS Frame {
+  Type (i) = TBD-00 (experiments use 0xff3e800),
+  Length (i),
+  Channel ID Lengths (8) ...,
```
Comment (GrumpyOldTroll) [inline draft-jholland-quic-multicast.md:235]: I think a list of { Length, ChannelID } structures is probably better.  There's an existing [WriteLengthPrefixedConnectionId](https://github.com/GrumpyOldTroll/quiche/blob/jake-demo-branch/quic/core/quic_data_writer.h#L88) function that can be used if it's structured that way.

Comment (MaxF12) [reply draft-jholland-quic-multicast.md:235]: Right, good point about the fragmentation and list format. I guess another option would be to just have a range of reserved channels, so something like

`Length start (8),
Start channel ID (8..160),
Length end (8),
End channel ID (8..160),'

I guess the question is if there is any serious reason why a server would want to have several entirely separate ranges of channel IDs? Or maybe you just allow to send several of those frames, in cases where it is really required?

**Comment by MaxF12** [comment]
Gave it a shot to split the properties frame as well now to maybe make it more straight forward and use the new frame as not only reservation of a Channel ID but also to communicate all the static properties of a channel (into two for now, splitting keys out could also come later)

File: draft-jholland-quic-multicast.md [hunk 2]
```diff
@@ -225,34 +225,71 @@ If a server or client somehow still detect a stateless reset for a channel, they
...
+      * **Author's Note:** I assume it’s not better to use a TLS CipherSuite because there is no KDF stage for deriving these keys (they are a strict server-to-client advertisement), so the Hash part would be unused? (<https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml#tls-parameters-4>)
+
+A client MUST NOT use the channel ID included in the MC_CHANNEL_ANNOUNCE frame as a connection ID for the unicast connection. If it is already in use, the client SHOULD retire it as soon as possible.
+As the server knows which connection IDs are in use by the client, it SHOULD wait with the sending of a MC_CHANNEL_JOIN frame until the channel ID associated with it has been retired by the client.
+
+As the properties in MC_CHANNEL_ANNOUNCE frames are immutable during the lifetime of a channel, a server SHOULD NOT send a MC_CHANNEL_ANNOUNCE frame for the same channel more than once to each client.
+
+A server MUST send an MC_CHANNEL_ANNOUNCE frame for a channel before sending a MC_CHANNEL_PROPERTIES or MC_CHANNEL_JOIN frame for it.
```
[the remark this answers is not in the document]
Comment (GrumpyOldTroll) [reply draft-jholland-quic-multicast.md:270]: Plus ~90 for the properties as a max size.  Hmm..., not many of these sets per packet.  I guess we probably do need the reservation if we're doing it this way.

File: draft-jholland-quic-multicast.md [hunk 3]
```diff
@@ -262,76 +299,44 @@ MC_CHANNEL_PROPERTIES Frame {
...
- * Max Streams: Present when Has Max Streams is set in the Content Field.  The maximum stream ID that might appear in the channel.  If a client joined to this channel can raise its Max Streams limit up to or above this value it SHOULD do so, otherwise it SHOULD leave or decline join for the channel with Max Streams Exceeded.
- * ACK Bundle Size: Present when Has Ack Bundle Size is set in the Content Field.  The minimum number of ACKs a client should send in a single QUIC packet. If the max_ack_delay would force a client to send a packet that only consists of MC_CHANNEL_ACK frames, it SHOULD instead wait with sending until at least the specified number of acknowledgements have been collected. However, the Client MUST send any pending acknowledgements at least once per Max Idle Time to prevent the Server from perceiving the channel as interrupted.
+The properties of a channel MAY change during its lifetime. As such, a server SHOULD wait with sending an MC_CHANNEL_PROPERTIES frame intended to provide the initial properties to a client that hasn't joined the channel yet until immediately before sending the MC_CHANNEL_JOIN frame.
```
Comment (GrumpyOldTroll) [inline draft-jholland-quic-multicast.md:332]: suggest replacing from "SHOULD ..." with: "SHOULD NOT send properties for channels except those the client has joined or will be imminently asked to join."

File: draft-jholland-quic-multicast.md [hunk 4]
```diff
@@ -262,76 +299,44 @@ MC_CHANNEL_PROPERTIES Frame {
...
 
 ## MC_CHANNEL_JOIN {#channel-join-frame}
 
 An MC_CHANNEL_JOIN frame (type TBD-02) is sent from server to client and requests that a client join the given transport addresses and ports and process packets with the given Channel ID according to the corresponding MC_CHANNEL_PROPERTIES.
 
+A client cannot join a multicast channel without first receiving a MC_CHANNEL_ANNOUNCE and MC_CHANNEL_PROPERTIES frame which together set all the values for a channel.
```
Comment (GrumpyOldTroll) [inline draft-jholland-quic-multicast.md:338]: Suggest adding:
If a client receives a MC_CHANNEL_JOIN for a channel for which it has not received both, it MUST respond with a MC_CLIENT_CHANNEL_STATE with State "Declined Join" and reason "Missing Properties".  The server MAY send another MC_CHANNEL_JOIN after retransmitting the MC_CHANNEL_PROPERTIES and receiving an acknowledgement indicating receipt of the MC_CHANNEL_ANNOUNCE.

**Pull request merged**